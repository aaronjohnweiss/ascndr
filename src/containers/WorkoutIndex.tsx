import React, { useState } from 'react'
import Button from 'react-bootstrap/Button'
import EntityModal from '../components/EntityModal'
import { isLoaded, useFirebase } from 'react-redux-firebase'
import Workout from '../components/Workout'
import { validateWorkoutFields, workoutFields } from '../templates/workoutFields'
import { getUser, useDatabase } from '../redux/selectors/selectors'

const WorkoutIndex = () => {
  const { uid } = getUser()
  const firebaseState = useDatabase()
  const workouts = firebaseState.workouts.getOrdered(['owner', uid])

  const firebase = useFirebase()

  const [showModal, setShowModal] = useState(false)

  const openModal = () => setShowModal(true)
  const closeModal = () => setShowModal(false)

  const addWorkout = workout => {
    firebase.push('workouts', { ...workout, uid, startTime: new Date().getTime() })
    closeModal()
  }

  if (!isLoaded(workouts)) return <>Loading</>

  workouts.sort((a, b) => b.value.startTime - a.value.startTime)

  return (
    <>
      <div className="d-grid d-block mb-4">
        <Button id="add-workout" variant="primary" onClick={openModal}>
          Add Workout
        </Button>
      </div>
      {workouts.map(workout => (
        <Workout workout={workout.value} key={workout.key} />
      ))}
      <EntityModal
        show={showModal}
        handleClose={closeModal}
        handleSubmit={addWorkout}
        fields={workoutFields}
        initialValues={{ intensity: 1 }}
        validateState={validateWorkoutFields}
      />
    </>
  )
}

export default WorkoutIndex
