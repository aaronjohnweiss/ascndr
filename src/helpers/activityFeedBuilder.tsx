import {OrderedList, Persisted} from "../types/Firebase";
import {Route, RouteVideo} from "../types/Route";
import {FinishedSession, isFinished, Session} from "../types/Session";
import {groupBy} from "./filterUtils";
import {calculateProjectTimes, isSent, Project} from "../components/RoutesIndex";
import {buildSort, contramap, sortBy} from "./sortBuilder";
import moment from "moment";
import {Workout} from "../types/Workout";
import {entries} from "./recordUtils";
import {Gym} from "../types/Gym";
import {User} from "../types/User";
import {Grade} from "../types/Grade";
import {calculateProgression} from "../components/GradeHistory";
import {toObj} from "./objectConverters";
import {ALL_STYLES} from "./gradeUtils";

export type AggregateSessionMilestone = {
    count: number,
    unit: 'sessionCount' | 'duration'
} | {
    grade: Grade,
    unit: 'grade'
}
type FeedData = {
    _type: 'session',
    value: Persisted<Session>
} | {
    _type: 'milestone',
    value: AggregateSessionMilestone
} | {
    _type: 'workout',
    value: Persisted<Workout>
} | {
    _type: 'video',
    value: {
        routeKey: string,
        video: RouteVideo
    }
} | {
    _type: 'project',
    value: {
        routeKey: string,
        project: Project
    }
}
export type FeedItem = {
    date: number,
    uid: string,
    link?: string,
    data: FeedData
}
type FeedItemType = FeedData['_type']
// Sort order - earlier items in this list will be displayed before later items, in case the items' dates are the same
const FeedItemTypes: FeedItemType[] = [
    'project',
    'milestone',
    'video',
    'session',
    'workout',
]

/**
 * Build a sorted list of activity feed items based on the provided data
 */
export const buildFeedData = (uid: string, gyms: OrderedList<Gym>, sessions: OrderedList<Session>, users: OrderedList<User>, routes: OrderedList<Route>, workouts: OrderedList<Workout>): FeedItem[] => {
    const feedData = [
        ...getSessionFeedItems(sessions),
        ...getSessionMilestoneFeedItems(sessions),
        ...getGradeMilestoneFeedItems(sessions, routes),
        ...getWorkoutFeedItems(workouts),
        ...getVideoFeedItems(routes, users),
        ...getProjectFeedItems(routes, sessions),
    ]

    return feedData.sort(buildSort(
        sortBy<FeedItem>()('date').descending,
        // Break ties according to the order of FeedItemTypes: earlier types in the array should come first
        sortBy<FeedItem>()('data', contramap(data => FeedItemTypes.indexOf(data._type))).ascending
    ))
}

/**
 * Simple feed item for each session
 */
const getSessionFeedItems = (sessions: OrderedList<Session>): FeedItem[] =>
    sessions.map(session => ({
        date: session.value.startTime,
        uid: session.value.uid,
        link: `/sessions/${session.key}`,
        data: {
            _type: 'session',
            value: session
        }
    }))

/**
 * Simple feed item for each workout
 */
const getWorkoutFeedItems = (workouts: OrderedList<Workout>): FeedItem[] =>
    workouts.map(workout => ({
        date: workout.value.startTime,
        uid: workout.value.uid,
        data: {
            _type: 'workout',
            value: workout
        }
    }))

/**
 * Aggregated session milestones (total hours climbed, total sessions climbed) per user
 */
const getSessionMilestoneFeedItems = (sessions: OrderedList<Session>): FeedItem[] => {
    const sessionsByUser = groupBy(sessions, 'uid')
    return entries(sessionsByUser).flatMap(([uid, userSessions]) => {
        const milestones = getSessionCountMilestones(userSessions)
        return milestones.map((milestone) => ({
            date: milestone.date,
            uid,
            data: {
                _type: 'milestone',
                value: milestone
            }
        }))
    })
}

/**
 * New highest grades per user
 */
const getGradeMilestoneFeedItems = (sessions: OrderedList<Session>, routes: OrderedList<Route>): FeedItem[] => {
    const sessionsByUser = groupBy(sessions, 'uid')
    return entries(sessionsByUser).flatMap(([uid, userSessions]) => {
        // For each user, calculate their full progression history in all styles. This ignores hiatuses.
        const milestones = calculateProgression(userSessions, toObj(routes), true, [...ALL_STYLES], false)
        return milestones.flatMap(({firsts}) => firsts).map(milestone => ({
            date: milestone.date,
            uid,
            link: `/sessions/${milestone.key}`,
            data: {
                _type: 'milestone',
                value: {
                    unit: 'grade',
                    grade: milestone.grade
                }
            }
        }))
    })
}

/**
 * Video uploads
 */
const getVideoFeedItems = (routes: OrderedList<Route>, users: OrderedList<User>): FeedItem[] => {
    return routes.flatMap(({key, value}) => value.videos
        ?.filter(video => users.some(u => u.value.uid === video.uid))
        .map(video => ({
            date: video.date,
            uid: video.uid,
            data: {
                _type: 'video',
                value: {
                    routeKey: key,
                    video,
                }
            }
        })) || [])
}

/**
 * Minimum number of sessions for a project to show up in the feed
 */
const MIN_PROJECT_SESSIONS = 2
/**
 * Projects sent by each user (routes that took at least {@link MIN_PROJECT_SESSIONS} sessions to climb successfully)
 */
const getProjectFeedItems = (routes: OrderedList<Route>, sessions: OrderedList<Session>): FeedItem[] => {
    const sessionsByUser = groupBy(sessions, 'uid')

    return routes.flatMap(route => calculateProjectTimes(route.key, sessionsByUser)
        // Exclude in-progress projects
        .filter(isSent)
        // Exclude projects below session count threshold
        .filter(project => project.sessionCount >= MIN_PROJECT_SESSIONS)
        .map(project => ({
            date: project.sentDate,
            uid: project.uid,
            link: `/routes/${route.key}`,
            data: {
                _type: 'project',
                value: {
                    routeKey: route.key,
                    project
                }
            }
        })))
}

/**
 * Milestone counts for aggregated session info
 */
const lowMilestones = [1, 5, 10, 25, 50, 75]
/**
 * Determine if this is a milestone session count (any entry of {@link lowMilestones} or any multiple of 100)
 */
const isMilestoneCount = (n: number) => lowMilestones.some(milestone => n === milestone) || n % 100 === 0

/**
 * Return any milestone duration passed in this session (any session that brings the total above {@link lowMilestones} or any multiple of 100 hrs climbed)
 */
const getMilestoneDuration = (before: number, after: number): number | undefined => {
    const lowMilestone = lowMilestones.findLast(milestone => before < milestone && milestone <= after);
    if (lowMilestone !== undefined) return lowMilestone

    if (after < 100) return undefined

    const beforeHundred = Math.floor(before / 100)
    const afterHundred = Math.floor(after / 100)

    if (beforeHundred === afterHundred) return undefined

    return afterHundred * 100
}


type SessionCountMilestone = AggregateSessionMilestone & {
    date: number
}
/**
 * Get milestones based off of the provided sessions. This will not slice up the provided sessions, unlike {@link getSessionMilestoneFeedItems} which slices by user.
 */
const getSessionCountMilestones = (sessions: OrderedList<Session>): SessionCountMilestone[] => {

    const sortedSessions = sessions.map(session => session.value)
        // Ignore in-progress sessions (milestones should come after a session is finishes)
        .filter(isFinished)
        // Sort by oldest first
        .sort(sortBy<FinishedSession>()('endTime').ascending)

    const milestones: SessionCountMilestone[] = []
    // Cumulative duration climbed
    let duration = 0
    for (let i = 0; i < sortedSessions.length; i++) {
        const session = sortedSessions[i]
        // Check session count for milestone
        if (isMilestoneCount(i + 1)) {
            milestones.push({count: i + 1, date: session.endTime, unit: 'sessionCount'})
        }
        // Check cumulative session duration for milestone
        const sessionDuration = moment(session.endTime).diff(moment(session.startTime), 'hours', true)
        const milestoneDuration = getMilestoneDuration(duration, duration + sessionDuration)
        if (milestoneDuration !== undefined) {
            milestones.push({
                count: milestoneDuration,
                date: session.endTime,
                unit: 'duration'
            })
        }
        // Update running total for duration
        duration = duration + sessionDuration
    }
    return milestones
}