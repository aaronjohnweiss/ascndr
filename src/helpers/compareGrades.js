const topropePrefix = '5.'
const compareGrades = (g1, g2) => {
    if (g1.indexOf(topropePrefix) === 0 && g2.indexOf(topropePrefix) === 0) {
        const g1Grade = parseInt(g1.substring(topropePrefix.length))
        const g2Grade = parseInt(g2.substring(topropePrefix.length))
        if (g1Grade > g2Grade) return 1
        if (g2Grade > g1Grade) return -1

        // Grades are the same number... differentiate by +/-
        if ((g1.endsWith('+') && !g2.endsWith('+')) || (g2.endsWith('-') && !g1.endsWith('-'))) return 1
        if ((g2.endsWith('+') && !g1.endsWith('+')) || (g1.endsWith('-') && !g2.endsWith('-'))) return -1

        return 0
    }

    return g1.localeCompare(g2)
}

export default compareGrades