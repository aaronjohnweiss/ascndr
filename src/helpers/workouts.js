export const WORKOUT_CATEGORIES = [
    'Strength',
    'Flexibility',
    'Hangboard',
    'Yoga'
]

export const MAX_INTENSITY = 4;
