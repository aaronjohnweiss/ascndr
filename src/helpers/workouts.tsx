export const WORKOUT_CATEGORIES = ['Strength', 'Flexibility', 'Hangboard', 'Yoga'] as const

export const MAX_INTENSITY = 4
