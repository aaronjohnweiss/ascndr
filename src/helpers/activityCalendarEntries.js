import moment from "moment";
import {dateString} from "./dateUtils";
import {routeCountForSession} from "../components/StatsIndex";
import {ALL_STYLES} from "./gradeUtils";
import {percentile} from "./mathUtils";
import {WORKOUT_CATEGORIES} from "./workouts";
import {CalendarMode, getPreferences} from "../components/ActivityCalendarSettingsModal";
import {findUser, getSessionsForUser, getUserName, getWorkoutsForUser} from "./filterUtils";

export const getEmptyCounts = cutoffDate => {
    const tmpDate = cutoffDate.clone();
    const now = moment();
    const emptyCounts = {};
    while (tmpDate.isBefore(now)) {
        emptyCounts[dateString(tmpDate.valueOf())] = {count: 0, level: 0};
        tmpDate.add(1, 'day');
    }
    return emptyCounts;
}

export const reduceCount = (acc = 0, count = 0) => count + acc
export const reduceLevel = (acc = 0, level = 0) => Math.max(level, acc)

export const buildCounts = ({data, getTime = x => x.startTime, getCount = () => 1, getLevel = () => 0, cutoffDate}) =>
    data.filter(x => cutoffDate.isBefore(moment(getTime(x))))
        .map(x => ({
            date: dateString(getTime(x)),
            count: getCount(x),
            level: getLevel(x),
        }))
        .reduce((acc, entry) => ({
            ...acc,
            [entry.date]: {
                count: reduceCount(acc[entry.date].count, entry.count),
                level: reduceLevel(acc[entry.date].level, entry.level)
            }
        }), getEmptyCounts(cutoffDate));

export const getLevelByPercentile = (data) => {
    const allCounts = data.map(x => x.count).filter(val => val > 0).sort((a, b) => a - b);
    const q25 = percentile(allCounts, 0.25);
    const q50 = percentile(allCounts, 0.50);
    const q75 = percentile(allCounts, 0.75);

    return count => {
        if (count === 0) return 0;
        if (count <= q25) return 1;
        if (count <= q50) return 2;
        if (count <= q75) return 3;
        return 4;
    }
}

const buildArray = map => Object.entries(map)
    .map(([date, {count, level}]) => ({date, count, level}))
    .sort((a, b) => a.date.localeCompare(b.date));

const getData = (ctx) => buildArray(buildCounts(ctx));

const getDataWithPercentiles = (ctx) => {
    const data = getData(ctx)

    const getLevel = getLevelByPercentile(data)

    return data.map(x => ({
        ...x,
        level: getLevel(x.count)
    }));
}

const excludeEmpty = data => data.filter(x => x.data.some(day => day.level > 0))

const mergeData = ({data, label}) => {
    const dataByDay = data.map(x => x.data).flat().reduce((acc, entry) => ({
        ...acc,
        [entry.date]: {
            count: reduceCount(acc[entry.date] && acc[entry.date].count, entry.count),
            level: reduceLevel(acc[entry.date] && acc[entry.date].level, entry.level)
        }
    }), {});

    return [{label, data: buildArray(dataByDay)}];
}

export const getCalendarData = ({user, users, sessions, routes, workouts, cutoffDate}) => {
    const preferences = getPreferences(user)

    if (preferences.mode === CalendarMode.FRIENDS) {
        const usersForCalendar = [user, ...preferences.friends.map(uid => findUser(users, uid))]
        const userData = usersForCalendar
            .map(u => mergeData({data: getUserData({user: u, preferences, sessions, routes, workouts, cutoffDate}), label: getUserName(u)}))
            .flat()
        return excludeEmpty(userData)
    } else {
        return excludeEmpty(getUserData({user, preferences, sessions, routes, workouts, cutoffDate}))
    }
}

export const getUserData = ({user, preferences, sessions, routes, workouts, cutoffDate}) => [
    ...getSessionData({sessions: getSessionsForUser(sessions, user.uid), routes, cutoffDate}),
    ...getWorkoutData({workouts: getWorkoutsForUser(workouts,  user.uid), cutoffDate, preferences}),
]

export const getSessionData = ({sessions, routes, cutoffDate}) => {
    const sessionData = getDataWithPercentiles({
        data: sessions.map(x => x.value),
        cutoffDate,
        getCount: session => routeCountForSession(session, routes, ALL_STYLES, true)
    })

    return [{label: 'Climbing', data: sessionData}]
}

export const getWorkoutData = ({workouts, cutoffDate, preferences}) => {
    if (!preferences.includeWorkouts) return []

    const allWorkoutData = excludeEmpty(WORKOUT_CATEGORIES.map(category => ({
        label: category,
        data: getData({
            data: workouts.map(x => x.value).filter(workout => workout.categories.includes(category)),
            cutoffDate,
            getLevel: workout => workout.intensity
        })
    })))

    if (!preferences.splitWorkouts) return mergeData({data: allWorkoutData, label: 'Workout'})
    return allWorkoutData
}