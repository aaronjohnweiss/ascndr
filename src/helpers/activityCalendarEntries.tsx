import moment from "moment";
import {dateString} from "./dateUtils";
import {routeCountForSession} from "../components/StatsIndex";
import {ALL_STYLES} from "./gradeUtils";
import {percentile} from "./mathUtils";
import {WORKOUT_CATEGORIES} from "./workouts";
import {getPreferences} from "../components/ActivityCalendarSettingsModal";
import {findUser, getSessionsForUser, getUserName, getWorkoutsForUser} from "./filterUtils";
import {ActivityCalendarPreferences, CalendarMode, User} from "../types/User";
import {Data, OrderedList} from "../types/Firebase";
import {Session} from "../types/Session";
import {Route} from "../types/Route";
import {Workout} from "../types/Workout";

interface Count {
    count: number,
    level: number,
}

export interface CountForDate extends Count {
    date: string
}

export interface MultiCountForDate {
    date: string
    counts: number[]
    levels: number[]
}

export interface LabeledData {
    label: string,
    data: CountForDate[]
}
export const getEmptyCounts = (cutoffDate: moment.Moment) => {
    const tmpDate = cutoffDate.clone();
    const now = moment();
    const emptyCounts: Record<string, Count> = {};
    while (tmpDate.isBefore(now)) {
        emptyCounts[dateString(tmpDate.valueOf())] = {count: 0, level: 0};
        tmpDate.add(1, 'day');
    }
    return emptyCounts;
}

export const reduceCount = (acc = 0, count = 0) => count + acc
export const reduceLevel = (acc = 0, level = 0) => Math.max(level, acc)

interface BuildCountCtx<T> {
    data: T[]
    getTime: (x: T) => number,
    getCount?: (x: T) => number,
    getLevel?: (x: T) => number,
    cutoffDate: moment.Moment
}
export const buildCounts = <T,>({data, getTime, getCount = () => 1, getLevel = () => 0, cutoffDate}: BuildCountCtx<T>): Record<string, Count> =>
    data.filter(x => cutoffDate.isBefore(moment(getTime(x))))
        .map(x => ({
            date: dateString(getTime(x)),
            count: getCount(x),
            level: getLevel(x),
        }))
        .reduce((acc, entry) => ({
            ...acc,
            [entry.date]: {
                count: reduceCount(acc[entry.date].count, entry.count),
                level: reduceLevel(acc[entry.date].level, entry.level)
            }
        }), getEmptyCounts(cutoffDate));

export const getLevelByPercentile = (data: CountForDate[]) => {
    const allCounts = data.map(x => x.count).filter(val => val > 0).sort((a, b) => a - b);
    const q25 = percentile(allCounts, 0.25);
    const q50 = percentile(allCounts, 0.50);
    const q75 = percentile(allCounts, 0.75);

    return (count: number) => {
        if (count === 0) return 0;
        if (count <= q25) return 1;
        if (count <= q50) return 2;
        if (count <= q75) return 3;
        return 4;
    }
}

const buildArray = (map: Record<string, Count>): CountForDate[] => Object.entries(map)
    .map(([date, {count, level}]) => ({date, count, level}))
    .sort((a, b) => a.date.localeCompare(b.date));

const getData = <T,> (ctx: BuildCountCtx<T>): CountForDate[] => buildArray(buildCounts(ctx));

const getDataWithPercentiles = <T,> (ctx: BuildCountCtx<T>) => {
    const data = getData(ctx)

    const getLevel = getLevelByPercentile(data)

    return data.map(x => ({
        ...x,
        level: getLevel(x.count)
    }));
}

const excludeEmpty = (arr: LabeledData[]) => arr.filter(x => x.data.some(day => day.level > 0))

const mergeData = ({data, label}: {data: LabeledData[], label: string}): LabeledData[] => {
    const dataByDay = data.map(x => x.data).flat().reduce((acc, entry) => ({
        ...acc,
        [entry.date]: {
            count: reduceCount(acc[entry.date] && acc[entry.date].count, entry.count),
            level: reduceLevel(acc[entry.date] && acc[entry.date].level, entry.level)
        }
    }), {} as Record<string, Count>);

    return [{label, data: buildArray(dataByDay)}];
}

export interface Ctx {
    user: User
    users: OrderedList<User>
    sessions: OrderedList<Session>
    routes: Data<Route>
    workouts: OrderedList<Workout>
    cutoffDate: moment.Moment
    preferences: ActivityCalendarPreferences
}

export const getCalendarData = ({user, users, sessions, routes, workouts, cutoffDate}: Omit<Ctx, 'preferences'>): LabeledData[] => {
    const preferences = getPreferences(user)

    if (preferences.mode === CalendarMode.FRIENDS) {
        const usersForCalendar = [user, ...preferences.friends.map(uid => findUser(users, uid))]
        const userData = usersForCalendar
            .map(u => mergeData({data: getUserData({user: u, preferences, sessions, routes, workouts, cutoffDate}), label: getUserName(u)}))
            .flat()
        return excludeEmpty(userData)
    } else {
        return excludeEmpty(getUserData({user, preferences, sessions, routes, workouts, cutoffDate}))
    }
}

export const getUserData = ({user, preferences, sessions, routes, workouts, cutoffDate}: Pick<Ctx, 'user' | 'preferences' | 'sessions' | 'routes' | 'workouts' | 'cutoffDate'>) => [
    ...getSessionData({sessions: getSessionsForUser(sessions, user.uid), routes, cutoffDate}),
    ...getWorkoutData({workouts: getWorkoutsForUser(workouts,  user.uid), cutoffDate, preferences}),
]

export const getSessionData = ({sessions, routes, cutoffDate}: Pick<Ctx, 'sessions' | 'routes' | 'cutoffDate'>) => {
    const sessionData = getDataWithPercentiles({
        data: sessions.map(x => x.value),
        cutoffDate,
        getTime: session => session.startTime,
        getCount: session => routeCountForSession(session, routes, [...ALL_STYLES], true)
    })

    return [{label: 'Climbing', data: sessionData}]
}

export const getWorkoutData = ({workouts, cutoffDate, preferences}: Pick<Ctx, 'workouts' | 'cutoffDate' | 'preferences'>) => {
    if (!preferences.includeWorkouts) return []

    const allWorkoutData = excludeEmpty(WORKOUT_CATEGORIES.map(category => ({
        label: category,
        data: getData({
            data: workouts.map(x => x.value).filter(workout => workout.categories.includes(category)),
            cutoffDate,
            getTime: workout => workout.startTime,
            getLevel: workout => workout.intensity,
        })
    })))

    if (!preferences.splitWorkouts) return mergeData({data: allWorkoutData, label: 'Workout'})
    return allWorkoutData
}