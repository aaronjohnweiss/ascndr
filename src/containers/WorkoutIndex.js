import React, {useState} from 'react'
import {compose} from 'redux'
import {connect} from 'react-redux'
import Button from 'react-bootstrap/Button'
import EntityModal from '../components/EntityModal'
import {firebaseConnect, isLoaded} from 'react-redux-firebase'
import {getWorkoutsForUser} from '../helpers/filterUtils';
import Workout from "../components/Workout";
import {validateWorkoutFields, workoutFields} from "../templates/workoutFields";

const WorkoutIndex = ({auth: {uid}, workouts, firebase}) => {

    const [showModal, setShowModal] = useState(false)

    const openModal = () => setShowModal(true)
    const closeModal = () => setShowModal(false)

    const addWorkout = (workout) => {
        firebase.push('workouts', { ...workout, uid, startTime: new Date().getTime()})
        closeModal()
    }

    if (!isLoaded(workouts)) return 'Loading'

    const workoutsForUser = getWorkoutsForUser(workouts, uid)
    workoutsForUser.sort((a, b) => b.value.startTime - a.value.startTime)

    return (
        <>
            <div className="d-grid d-block mb-4">
                <Button variant='primary' onClick={openModal}>
                    Add Workout
                </Button>
            </div>
            {workoutsForUser.map((workout) => <Workout workout={workout.value} key={workout.key} />)}
            <EntityModal show={showModal}
                         handleClose={closeModal}
                         handleSubmit={addWorkout}
                         fields={workoutFields}
                         initialValues={{intensity: 1}}
                         validateState={validateWorkoutFields}
            />
        </>
    )
}

const mapStateToProps = state => {
    return {
        workouts: state.firebase.ordered.workouts,
        auth: state.auth
    }
}

export default compose(
    firebaseConnect([
        {path: 'workouts'},
    ]),
    connect(mapStateToProps)
)(WorkoutIndex)
