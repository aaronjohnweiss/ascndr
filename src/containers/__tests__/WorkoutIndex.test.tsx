import {
  mockFirebase,
  mountWithProviders,
  setupMockFirebase,
  user,
} from '../../redux/__tests__/reducer.test'
import WorkoutIndex from '../WorkoutIndex'
import { WORKOUT_CATEGORIES } from '../../helpers/workouts'

describe('WorkoutIndex', () => {
  beforeEach(() => {
    setupMockFirebase()
  })

  it('should render', () => {
    expect(mountWithProviders(WorkoutIndex).component).toMatchSnapshot()
  })

  it('should allow adding a workout', () => {
    const wrapper = mountWithProviders(WorkoutIndex).wrapper

    const intensity = 3
    const categories = WORKOUT_CATEGORIES.filter((category, idx) => idx % 2 === 0)

    wrapper.find('Button#add-workout').simulate('click')
    const modal = wrapper.find('EntityModal')
    modal.find('IntensityPicker').invoke('onChange')(intensity)
    modal.find('CategoryPicker').invoke('onChange')(categories)
    modal.find('Button#modal-submit').simulate('click')

    expect(mockFirebase.push).toHaveBeenCalledWith(
      `workouts`,
      expect.objectContaining({
        uid: user.uid,
        intensity,
        categories,
      }),
    )
  })
})
