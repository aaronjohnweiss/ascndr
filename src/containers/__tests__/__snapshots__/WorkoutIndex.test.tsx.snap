// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkoutIndex should render 1`] = `
<WorkoutIndex
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "canGo": [Function],
      "createHref": [Function],
      "entries": Array [
        Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
      ],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "index": 0,
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
  location={
    Object {
      "hash": "",
      "pathname": "/",
      "search": "",
      "state": undefined,
    }
  }
  match={
    Object {
      "isExact": true,
      "params": Object {},
      "path": "/",
      "url": "/",
    }
  }
>
  <div
    className="d-grid d-block mb-4"
  >
    <Button
      active={false}
      disabled={false}
      id="add-workout"
      onClick={[Function]}
      variant="primary"
    >
      <button
        className="btn btn-primary"
        disabled={false}
        id="add-workout"
        onClick={[Function]}
        type="button"
      >
        Add Workout
      </button>
    </Button>
  </div>
  <Workout
    key="workoutA"
    workout={
      Object {
        "_type": "workout",
        "categories": Array [],
        "intensity": 0,
        "startTime": 1671926400000,
        "uid": "myUid",
      }
    }
  >
    <Card
      body={false}
    >
      <div
        className="card"
      >
        <CardBody>
          <div
            className="card-body"
          >
            <CardTitle>
              <ForwardRef
                className="card-title"
              >
                <div
                  className="card-title h5"
                >
                  Sat Dec 24 2022
                </div>
              </ForwardRef>
            </CardTitle>
            <CardSubtitle
              className="mb-2 text-muted"
            >
              <ForwardRef
                className="mb-2 text-muted card-subtitle"
              >
                <div
                  className="mb-2 text-muted card-subtitle h6"
                >
                  Intensity: 
                  0
                </div>
              </ForwardRef>
            </CardSubtitle>
            <CardText>
              <p
                className="card-text"
              />
            </CardText>
          </div>
        </CardBody>
      </div>
    </Card>
  </Workout>
  <EntityModal
    fields={
      Array [
        Object {
          "name": "intensity",
          "options": Object {
            "component": [Function],
            "type": "custom",
          },
          "title": "Intensity",
        },
        Object {
          "name": "categories",
          "options": Object {
            "component": [Function],
            "type": "custom",
          },
          "title": "Categories",
        },
      ]
    }
    handleClose={[Function]}
    handleSubmit={[Function]}
    initialValues={
      Object {
        "intensity": 1,
      }
    }
    show={false}
    validateState={[Function]}
  >
    <Modal
      animation={true}
      autoFocus={true}
      backdrop={true}
      dialogAs={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "displayName": "ModalDialog",
          "render": [Function],
        }
      }
      enforceFocus={true}
      keyboard={true}
      onHide={[Function]}
      restoreFocus={true}
      show={false}
    >
      <Modal
        autoFocus={true}
        backdrop={true}
        backdropTransition={[Function]}
        containerClassName="modal-open"
        enforceFocus={true}
        keyboard={true}
        manager={
          BootstrapModalManager {
            "containers": Array [],
            "data": Array [],
            "handleContainerOverflow": true,
            "hideSiblingNodes": true,
            "modals": Array [],
            "scrollbarSize": 0,
          }
        }
        onEnter={[Function]}
        onEntering={[Function]}
        onEscapeKeyDown={[Function]}
        onExit={[Function]}
        onExited={[Function]}
        onHide={[Function]}
        renderBackdrop={[Function]}
        renderDialog={[Function]}
        restoreFocus={true}
        show={false}
        transition={[Function]}
      />
    </Modal>
    <ConfirmCancelModal
      handleConfirm={[Function]}
      hideModal={[Function]}
      modalTitle="Delete session?"
      showModal={false}
    >
      <Modal
        animation={true}
        autoFocus={true}
        backdrop={true}
        dialogAs={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "displayName": "ModalDialog",
            "render": [Function],
          }
        }
        enforceFocus={true}
        keyboard={true}
        onHide={[Function]}
        restoreFocus={true}
        show={false}
      >
        <Modal
          autoFocus={true}
          backdrop={true}
          backdropTransition={[Function]}
          containerClassName="modal-open"
          enforceFocus={true}
          keyboard={true}
          manager={
            BootstrapModalManager {
              "containers": Array [],
              "data": Array [],
              "handleContainerOverflow": true,
              "hideSiblingNodes": true,
              "modals": Array [],
              "scrollbarSize": 0,
            }
          }
          onEnter={[Function]}
          onEntering={[Function]}
          onEscapeKeyDown={[Function]}
          onExit={[Function]}
          onExited={[Function]}
          onHide={[Function]}
          renderBackdrop={[Function]}
          renderDialog={[Function]}
          restoreFocus={true}
          show={false}
          transition={[Function]}
        />
      </Modal>
    </ConfirmCancelModal>
  </EntityModal>
</WorkoutIndex>
`;
